magefiles/run.go

```diff
@@ -0,0 +1,209 @@
+	log.Println("Retry vagrant up if the first time the Windows node failed")
+	log.Println("##########################################################")
+
+	const maxAttempts = 10 // TODO: How many to allow? Make configurable via variables.yaml?
+	for i := 1; i < maxAttempts; i++ {
+		log.Printf("vagrant status winw1 - attempt %d of %d", i, maxAttempts)
+		output, err := sh.Output("vagrant", "status", "winw1")
+		if err != nil {
+			return err
+		}
+		status := extractMachineStatus(output, "winw1")
+		log.Println("winw1", status)
+		if strings.Contains(status, "running") {
+			break
+		}
+		log.Printf("vagrant up winw1 - attempt %d of %d", i, maxAttempts)
+		err = sh.Run("vagrant", "up", "winw1")
+		if err != nil {
+			return err
+		}
+	}
+
+	mustSetVagrantPrivateKeyPermissions()
+
+	for i := 1; i < maxAttempts; i++ {
+		log.Printf("kubectl get nodes | grep winw1  - attempt %d of %d", i, maxAttempts)
+		output, err := sh.Output("vagrant", "ssh", "controlplane", "-c", "kubectl get nodes")
+		if err != nil {
+			return err
+		}
+		log.Println(output)
+		if strings.Contains(output, "winw1") {
+			break
+		}
+		log.Printf("vagrant provision winw1 - attempt %d of %d", i, maxAttempts)
+		err = sh.Run("vagrant", "provision", "winw1")
+		if err != nil {
+			return err
+		}
+	}
+
+	// Cluster status //////////////////////////////////////////////////////////
+	joinedFile := filepath.Join(".lock", "joined")
+	log.Printf("Creating %s indicator file for Vagrantfile", joinedFile)
+	mustTouchFile(joinedFile)
+
+	cniFile := filepath.Join(".lock", "cni")
+	log.Printf("Creating %s indicator file for Vagrantfile", cniFile)
+	mustTouchFile(cniFile)
+
+	log.Println("Cluster created")
+	err = sh.Run("vagrant", "status")
+	if err != nil {
+		return err
+	}
+	err = sh.Run("vagrant", "ssh", "controlplane", "-c", "kubectl get nodes")
+	if err != nil {
+		return err
+	}
+
+	logTargetRunTime("Run")
+	return nil
+}
+
+func extractMachineStatus(statusOutput string, machineName string) string {
+	r := fmt.Sprintf(`(?m)^%s.*`, machineName)
+	m := regexp.MustCompile(r)
+	s := m.FindString(statusOutput)
+	v := strings.Split(s, machineName)
+	return strings.TrimSpace(v[1])
+}
+
+func mustTouchFile(filePath string) {
+	file, err := os.OpenFile(filePath, os.O_CREATE|os.O_RDWR|os.O_TRUNC, 644)
+	if err != nil {
+		panic(err)
+	}
+	err = file.Close()
+	if err != nil {
+		panic(err)
+	}
+}
+
+func mustSetVagrantPrivateKeyPermissions() {
+	var keyFiles = [...]string{
+		filepath.Join(".vagrant", "machines", "controlplane", "virtualbox", "private_key"),
+		filepath.Join(".vagrant", "machines", "winw1", "virtualbox", "private_key"),
+	}
+
+	for _, keyFile := range keyFiles {
+		_, err := os.Stat(filepath.Clean(keyFile))
+		if os.IsNotExist(err) {
+			continue // key file may not be created yet, skip for now
+		}
+		log.Printf("Setting SSH private key permissions for %s", keyFile)
+		if runtime.GOOS != "windows" {
+			err = os.Chmod(keyFile, 600)
+			if err != nil {
+				panic(err)
+			}
```

Comment:
nit: we could add a ``continue`` here, so we wouldn't need to write else for the block below.

Reply:
...old habits, but I've recently learned that in Go limiting number of branch levels should be preferred indeed.
I'll add `continue`.